File: C++/modular-exponentiation.cpp
#include <iostream>

///(a*b)%m = [(a%m)*(b*m)]%m

using namespace std;

long long modularExp(int a,long b,int m){

    if(b==0)return 1;

    if(b%2==0)  //even power
    {
        long long y = modularExp(a,b/2,m);
        return (y*y)%m;
    }
    //else odd
    return ((a%m)*modularExp(a,b-1,m))%m;

}

int main()
{

    int t;
    cin>>t;
    while(t-- > 0){

        int a;
        long long b;
        cin>>a>>b;

        long long ans = modularExp(a,b,10);
        cout<<ans<<endl;

    }

}

File: Java/modularExponentiation.java

public class modularExponentiation {
		
		  static int power(int x, int y, int p) 
		    {  
		        int res = 1;      
		          
		        x = x % p;  
		      
		        while (y > 0) 
		        { 
		            if((y & 1)==1) 
		                res = (res * x) % p; 
		            y = y >> 1;  
		            x = (x * x) % p;  
		        } 
		        return res; 
		    } 
		  
		    public static void main(String args[]) 
		    { 
		        int x = 10; 
		        int y = 100000; 
		        int m = 1000000007; 
		        // Gives ans as x^y%m
		        System.out.println("Power is " + power(x, y, m)); 
		    } 
		

	}


